runtime/server.go

```diff
@@ -0,0 +1,538 @@
+// Copyright 2016 The OPA Authors.  All rights reserved.
+// Use of this source code is governed by an Apache2
+// license that can be found in the LICENSE file.
+
+package runtime
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/gorilla/mux"
+	"github.com/open-policy-agent/opa/ast"
+	"github.com/open-policy-agent/opa/eval"
+	"github.com/open-policy-agent/opa/version"
+	"github.com/pkg/errors"
+)
+
+// apiErrorV1 models an error response sent to the client.
+type apiErrorV1 struct {
+	Code    int
+	Message string
+}
+
+func (err *apiErrorV1) Bytes() []byte {
+	if bs, err := json.MarshalIndent(err, "", "  "); err == nil {
+		return bs
+	}
+	return nil
+}
+
+// patchV1 models a single patch operation against a document.
+type patchV1 struct {
+	Op    string      `json:"op"`
+	Path  string      `json:"path"`
+	Value interface{} `json:"value"`
+}
+
+// policyV1 models a policy module in OPA.
+type policyV1 struct {
+	ID     string
+	Module *ast.Module
+}
+
+func (p *policyV1) Equal(other *policyV1) bool {
+	return p.ID == other.ID && p.Module.Equal(other.Module)
+}
+
+// resultSetV1 models the result of an ad-hoc query.
+type resultSetV1 []map[string]interface{}
+
+// Server contains runtime state specific to the server-mode persona, e.g.,
+// HTTP router.
+//
+// Notes:
+//
+// - In the future, the HTTP routing could be factored into a separate module
+// relying on the server (which would remain RPC-based). For now, it's simpler
+// to keep the HTTP routing and backend implementation in one place.
+type Server struct {
+	Addr    string
+	Persist bool
+	Runtime *Runtime
+	Router  *mux.Router
+
+	mtx sync.RWMutex
+}
+
+// NewServer returns a new Server.
+func NewServer(rt *Runtime, addr string, persist bool) *Server {
+
+	s := &Server{
+		Addr:    addr,
+		Persist: persist,
+		Runtime: rt,
+		Router:  mux.NewRouter(),
+	}
+
+	s.registerHandlerV1("/data/{path:.+}", "GET", s.v1DataGet)
+	s.registerHandlerV1("/data/{path:.+}", "PATCH", s.v1DataPatch)
+	s.registerHandlerV1("/policies", "GET", s.v1PoliciesList)
+	s.registerHandlerV1("/policies/{id}", "DELETE", s.v1PoliciesDelete)
+	s.registerHandlerV1("/policies/{id}", "GET", s.v1PoliciesGet)
+	s.registerHandlerV1("/policies/{id}/raw", "GET", s.v1PoliciesRawGet)
+	s.registerHandlerV1("/policies/{id}", "PUT", s.v1PoliciesPut)
+	s.registerHandlerV1("/query", "GET", s.v1QueryGet)
+
```

Comment:
Swagger is an option in the future.

Before I implemented the server I surveyed Go HTTP frameworks and Swagger support. As far as Swagger support goes, there is not too much available.

The best project I found was go-swagger. This would allow us to define the API in JSON/YAML/etc. and then generate the server code (i.e., models, handler boilerplate, etc.) The main reason for avoiding go-swagger for now is the lack of community behind it. It seems only a single person is working on it at the moment.

At the same time, we will need to have reference docs for the APIs soon. I'm still thinking about how to handle this.

One option would be to generate the Swagger spec (i.e., swagger.json) from annotations in our code (comments, etc.). This is not as nice as generating the code from the spec, but many projects work like this. Unfortunately, the tooling for this is not great. github.com/yvasiyarov/swagger exists but it was abandoned. It's been picked up by the go-swagger maintainer, so perhaps it will see improvement.

I think for now, it's OK to maintain the server code by hand. Swagger is great if you have a large number of endpoints and models to maintain. E.g., if your API is very domain specific, this is often the case. In our case, the APIs are fairly generic so I think there's less to gain from generating client code, etc.